rust/src/treefile.rs

```diff
@@ -100,32 +117,173 @@ fn treefile_parse_stream<R: io::Read>(
     Ok(treefile)
 }
 
+/// Given a treefile filename and an architecture, parse it and also
+/// open its external files.
+fn treefile_parse<P: AsRef<Path>>(
+    filename: P,
+    arch: Option<&str>,
+) -> io::Result<ConfigAndExternals> {
+    let filename = filename.as_ref();
+    let mut f = io::BufReader::new(fs::File::open(filename)?);
+    let basename = filename
+        .file_name()
+        .map(|s| s.to_string_lossy())
+        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Expected a filename"))?;
+    let fmt = if basename.ends_with(".yaml") || basename.ends_with(".yml") {
+        InputFormat::YAML
+    } else {
+        InputFormat::JSON
+    };
+    let tf = treefile_parse_stream(fmt, &mut f, arch).map_err(|e| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("Parsing {}: {}", filename.to_string_lossy(), e.to_string()),
+        )
+    })?;
+    let postprocess_script = if let Some(ref postprocess) = tf.postprocess_script.as_ref() {
+        Some(fs::File::open(filename.with_file_name(postprocess))?)
+    } else {
+        None
+    };
+    let mut add_files: collections::HashMap<String, fs::File> = collections::HashMap::new();
+    if let Some(ref add_file_names) = tf.add_files.as_ref() {
+        for (name, _) in add_file_names.iter() {
+            add_files.insert(name.clone(), fs::File::open(filename.with_file_name(name))?);
+        }
+    }
+    Ok(ConfigAndExternals {
+        config: tf,
+        externals: TreefileExternals {
+            postprocess_script,
+            add_files,
+        },
+    })
+}
+
+/// Merge a "basic" or non-array field. The semantics originally defined for
+/// these are that first-one wins.
+fn merge_basic_field<T>(dest: &mut Option<T>, src: &mut Option<T>) {
+    if dest.is_some() {
+        return;
+    }
+    *dest = src.take()
+}
+
+/// Merge a vector field by appending. This semantic was originally designed for
+/// the `packages` key.
+fn merge_vec_field<T>(dest: &mut Option<Vec<T>>, src: &mut Option<Vec<T>>) {
+    if let Some(ref mut srcv) = src.take() {
+        let mut v = dest.take().map_or_else(|| vec![], |v| v);
+        v.append(srcv);
+        *dest = Some(v)
```

Comment:
Wait, this is actually slightly changing the semantics now. Before we were prepending fields from includes, but now we're appending, right?

Which I think could matter in some cases? E.g. just from a quick scan:
- `initramfs_args` is passed directly down to `dracut` and not handled by us
- `postprocess`, i.e. inlined script ordering
- `add-files`... although ahh right, we essentially use `NOREPLACE` here so it would error out anyway on the second try to write to the same destination.